terraform: update all providers and move nixos provider under terraform.withPlugins

kalbasit opened the pull request:
###### Motivation for this change


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

/cc @jgeerds @zimbatm @peterhoeg

kalbasit committed:
terraform: update all the providers

kalbasit committed:
terraform-provider-nixos: move the provider to the inside of terraform.withPlugins

kalbasit committed:
terraform: update-all: add support for parsing list of providers in providers.txt

kalbasit committed:
terraform: add myself as a maintainer

GrahamcOfBorg commented:
<!--REQUEST_ID=aa4aad8e-fef0-4482-9af4-dfbc0d08ae72-->
Failure on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.47407&attempt_id=c8c836bf-cb06-4cd3-997f-b08a5177960f)

Attempted: terraform

<details><summary>Partial log (click to expand)</summary><p>

```
github.com/hashicorp/terraform/vendor/github.com/hashicorp/go-plugin
github.com/hashicorp/terraform/vendor/github.com/hashicorp/go-getter
github.com/hashicorp/terraform/moduledeps
github.com/hashicorp/terraform/config
github.com/hashicorp/terraform/config/module
github.com/hashicorp/terraform/terraform
github.com/hashicorp/terraform/plugin
github.com/hashicorp/terraform/tools/terraform-bundle
builder for '/nix/store/x1vxybzxkd4wpwvww3s5lv27sw83pb5q-terraform-0.11.8.drv' failed with exit code 51
error: build of '/nix/store/x1vxybzxkd4wpwvww3s5lv27sw83pb5q-terraform-0.11.8.drv' failed
```
</p></details>

flokli commented:
@kalbasit moving "official" and "standalone" providers into the same namespace makes a lot of sense, thanks :-)

Could you also move the two remaining providers, `terraform-provider-ibm` and `terraform-provider-libvirt` in there?

The former one should be completely straightforward, for the latter one we would need to add a way to `overrideAttrs`, setting `(propagated)buildInputs` - either from `update-all`, or in `providers/default.nix`.

kalbasit commented:
@flokli I tried moving both providers but I hit two issues that I thought were out of scope for this PR:

- `terraform-provider-ibm` provides a go package downloaded from `github.com/IBM-Cloud/terraform-provider-ibm` but the package itself expects to be cloned at `github.com/terraform-providers/terraform-provider-ibm` so this needs a special case in https://github.com/NixOS/nixpkgs/blob/0ad81753c5bea10f4c0bda9afaf76c7952756ab2/pkgs/applications/networking/cluster/terraform/providers/default.nix#L9
- `terraform-provider-libvirt` requires the `cdrtools` `propagatedBuildInputs`.

I'm hoping to first merge this PR, and work on the support for custom attributes in a separate PR.

flokli commented:
@kalbasit separate PR is fine, too - just wanted to make sure these don't slip through :-)

kalbasit commented:
@flokli It won't! I'll figure out the best way and submit a new PR sometime this week.

Mic92 commented on pkgs/applications/networking/cluster/terraform/providers/providers.txt:
```diff
@@ -0,0 +1,13 @@
+#  <organisation>/<repo>                                         - include only the named repository.
+
+# include all terraform-providers
+terraform-providers terraform-provider- terraform-provider-azure-classic\\|terraform-provider-scaffolding
```
Looks good, but you also need to filter out: terraform-provider-google-beta

```
*** terraform-provider-google-beta ***
fetching https://api.github.com/repos/terraform-providers/terraform-provider-google-beta/git/refs/tags?per_page=100
##O#- #
curl: (22) The requested URL returned error: 404 Not Found
The curl command has failed. Attempt 1/30
```

kalbasit replied:
@Mic92 PTAL.

GrahamcOfBorg commented:
<!--REQUEST_ID=b1f2ed44-ca5e-44df-a9e2-2b6d7170c81c-->
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.47407&attempt_id=40a77bab-8a37-4bad-9e93-4b910d0247df)

Attempted: terraform

<details><summary>Partial log (click to expand)</summary><p>

```
these paths will be fetched (14.63 MiB download, 105.12 MiB unpacked):
  /nix/store/0vbhc7kk6r10va4p18zy0lg3wv3046sf-tzdata-2018e
  /nix/store/6jb53clmhjm43ih5zy5hr433bq64rahb-terraform-0.11.8-bin
  /nix/store/kmqiiwrvb04hqnxhpjxl2jkn3md19pp2-iana-etc-20180711
copying path '/nix/store/kmqiiwrvb04hqnxhpjxl2jkn3md19pp2-iana-etc-20180711' from 'https://cache.nixos.org'...
copying path '/nix/store/0vbhc7kk6r10va4p18zy0lg3wv3046sf-tzdata-2018e' from 'https://cache.nixos.org'...
copying path '/nix/store/6jb53clmhjm43ih5zy5hr433bq64rahb-terraform-0.11.8-bin' from 'https://cache.nixos.org'...
/nix/store/6jb53clmhjm43ih5zy5hr433bq64rahb-terraform-0.11.8-bin
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=b1f2ed44-ca5e-44df-a9e2-2b6d7170c81c-->
Failure on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.47407&attempt_id=559be42c-25ce-4df4-bd93-de2e7be32460)

Attempted: terraform

<details><summary>Partial log (click to expand)</summary><p>

```
github.com/hashicorp/terraform/vendor/github.com/hashicorp/hcl2/hcl/hclsyntax
github.com/hashicorp/terraform/vendor/google.golang.org/grpc/transport
github.com/hashicorp/terraform/vendor/google.golang.org/grpc
github.com/hashicorp/terraform/vendor/github.com/hashicorp/hcl2/hcl/json
github.com/hashicorp/terraform/vendor/github.com/hashicorp/hcl2/hclparse
github.com/hashicorp/terraform/vendor/google.golang.org/grpc/health/grpc_health_v1
github.com/hashicorp/terraform/vendor/google.golang.org/grpc/health
github.com/hashicorp/terraform/vendor/github.com/hashicorp/go-plugin
builder for '/nix/store/x1vxybzxkd4wpwvww3s5lv27sw83pb5q-terraform-0.11.8.drv' failed with exit code 50
error: build of '/nix/store/x1vxybzxkd4wpwvww3s5lv27sw83pb5q-terraform-0.11.8.drv' failed
```
</p></details>

zimbatm committed:
terraform: move providers to terraform-providers

Before, providers were only built indirectly. Since proviers don't
depend on terraform to build they can be moved into their own collection
of packages. This also has the advantage that they can be reached
directly using an attribute path (Eg: terraform-providers.nixos).

Co-authored-by: Wael Nasreddine <wael.nasreddine@gmail.com>

zimbatm commented:
@kalbasit just pushed a change to your branch to move the providers to their own `terraform-providers` group. I wanted to add an alias from terraform-provider-nixos and the target wasn't reachable.

For the remaining 2 plugins, since they have special requirements they could just be moved under that group with callPackage links so the generic builder doesn't have to be extended.

kalbasit commented:
@zimbatm thx for the contribution but the eval is failing, can you take another look at it?

kalbasit committed:
terraform-providers: move ibm and libvirt underneath the terraform-providers

kalbasit commented:
@zimbatm never mind, I managed to fix it. The issue was that `recurseIntoAttrs` set up the packages as well as the functions `override`, `overrideDerivation` and the `recurseForDerivations` boolean. Removing them fixed the issue.

GrahamcOfBorg commented:
<!--REQUEST_ID=8ca426b1-1e78-4a29-983a-4ae10340e580-->
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.47407&attempt_id=3e093fde-abc1-4f14-97b7-65af210c4742)

Attempted: terraform

<details><summary>Partial log (click to expand)</summary><p>

```
these paths will be fetched (14.41 MiB download, 102.57 MiB unpacked):
  /nix/store/6jb53clmhjm43ih5zy5hr433bq64rahb-terraform-0.11.8-bin
copying path '/nix/store/6jb53clmhjm43ih5zy5hr433bq64rahb-terraform-0.11.8-bin' from 'https://cache.nixos.org'...
/nix/store/6jb53clmhjm43ih5zy5hr433bq64rahb-terraform-0.11.8-bin
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=8ca426b1-1e78-4a29-983a-4ae10340e580-->
Failure on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.47407&attempt_id=236308e1-47b7-4c78-aa80-c8bd33741e7a)

Attempted: terraform

<details><summary>Partial log (click to expand)</summary><p>

```
github.com/hashicorp/terraform/vendor/github.com/aws/aws-sdk-go/aws/session
github.com/hashicorp/terraform/vendor/github.com/hashicorp/go-getter
github.com/hashicorp/terraform/config
github.com/hashicorp/terraform/config/module
github.com/hashicorp/terraform/tools/terraform-bundle
github.com/hashicorp/terraform/tools/terraform-bundle/e2etest
github.com/hashicorp/terraform/vendor/github.com/hashicorp/go-version
github.com/hashicorp/terraform/version
builder for '/nix/store/x1vxybzxkd4wpwvww3s5lv27sw83pb5q-terraform-0.11.8.drv' failed with exit code 53
error: build of '/nix/store/x1vxybzxkd4wpwvww3s5lv27sw83pb5q-terraform-0.11.8.drv' failed
```
</p></details>

zimbatm commented:
nice work @kalbasit

zimbatm commented:
ported to 18.09 with `git cherry-pick -x -m 1 5547a8b`
